[Backend] Scrape keyword data from the Google Search page

Goose97 opened the pull request:
Closes [[Backend] Scrape keyword data from the Google Search page](https://github.com/Goose97/google-scraper-ruby/issues/14)

## What happened 👀

- Implement Google search service
- Implement Google search result parse service
- Implement scrape service 

## Insight 📝

- Use Nokogiri for HTML parsing
- Use Faraday as HTTP client. It offers great utilities to perform mocking during testing

## Proof Of Work 📹

All tests are passed

<img width="1512" alt="Screenshot 2023-07-27 at 21 22 23" src="https://github.com/Goose97/google-scraper-ruby/assets/35915460/4a6ff586-a158-4293-a629-1620dc323944">
<img width="1512" alt="Screenshot 2023-07-27 at 21 22 48" src="https://github.com/Goose97/google-scraper-ruby/assets/35915460/1ca91672-df5f-4621-bbea-509e66e0b4b7">
<img width="1512" alt="Screenshot 2023-07-27 at 21 22 10" src="https://github.com/Goose97/google-scraper-ruby/assets/35915460/ce06139a-9b94-4539-9d4c-ef93b14ade9c">

Goose97 committed:
Create database migration and set up models

Goose97 committed:
Implement keyword scraping command

Goose97 committed:
Extract errors to separate module

Goose97 committed:
Implement keyword search function

Goose97 committed:
Implement keyword parse function

Goose97 commented on app/commands/scrape_command.rb:
```diff
@@ -0,0 +1,47 @@
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  # rubocop:disable Metrics/AbcSize
```
I'm not sure how to address these two lint rules. IMO, this method is simple enough to a human to parse/understand.

I would prefer not to extract the rescue part to separate functions as it fragments the method.

olivierobert replied:
The method is too long because of the double `rescue` block and logging statements. While it is readable, there is complexity in the error handling, that's the code smell that Rubocop is raising. Using `rescue` blocks should be limited to maximum one (these are not comparable to functional paradigms like `case/else` or `with/else` in Elixir). To achieve this, it is best to raise one type of error, not multiple.

Goose97 replied:
Fixed in b0713248018bbc87ba507fb4a5ffa388a807a3df

olivierobert commented on .rubocop.yml:
```diff
@@ -21,6 +21,9 @@ AllCops:
   Enabled: false
 
+Style/NegatedIf:
+  Enabled: false
```
Let's avoid making changes to Rubocop as it defeats the purpose of having shared code style conventions. Adapting the code to fit the team's code style is best.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/7818449712a3ebb39bbdc369b1e88555cdf9ef4c

olivierobert commented on .rubocop.yml:
```diff
@@ -50,3 +53,6 @@ RSpec/NestedGroups:
   Max: 10
+
+RSpec/ExampleLength:
+  Max: 10
```
Same here. It might a sign that tests might need to be broken down.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/7818449712a3ebb39bbdc369b1e88555cdf9ef4c

andyduong1920 replied:
@Goose97 please double check as the `RSpec/ExampleLength:` is still here while it should be removed by the [7818449](https://github.com/Goose97/google-scraper-ruby/commit/7818449712a3ebb39bbdc369b1e88555cdf9ef4c), is there something wrong in the git flow?

Goose97 replied:
This change was reverted based on [Oliver suggestions](https://github.com/Goose97/google-scraper-ruby/pull/20#discussion_r1280060741). Should've resolve this one, my bad.

olivierobert commented on Gemfile:
```diff
@@ -11,6 +11,8 @@ gem 'discard' # Soft deletes for ActiveRecord
 gem 'bootsnap', require: false # Reduces boot times through caching; required in config/boot.rb
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby] # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
+gem 'faraday' # HTTP client
+gem 'nokogiri' # Gem to work with XML and HTML files
```
The descriptions can be improved slightly:

```suggestion
gem 'farday' # HTTP client library, with support for multiple backends
gem 'nokogiri' # Work with XML and HTML from Ruby
```

> **Note**
> In general, you can take the short description on the repository
> <img width="1294" alt="image" src="https://github.com/Goose97/google-scraper-ruby/assets/696529/0aae9a97-6c8e-4dd0-9254-822dfdf150a0">

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/802acb5ec77fc85f179ae4c144dd321054e4e1a5

olivierobert commented on app/commands/scrape_command.rb:
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+class ScrapeCommand
```
While there is nothing wrong with command objects, it is best to restrict yourself to the default objects listed in the project structure: https://nimblehq.co/compass/development/code-conventions/ruby/ruby-on-rails/#project-structure. The team's convention is to have only service objects, so if a service can call other objects, e.g., `ScrapService` can call the `Search` and `Parse` services. 

If the command pattern is something you want to introduce to the team, you can do it. You must follow [the change management guidelines](https://nimblehq.co/compass/intro/change-management/). It will likely more time though, due to the team sizes to get an agreement on this before we can merge this PR. So it is better to stick to the current conventions.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/8756759ccba2f72dc834b4e0b82d835307bbacda

olivierobert commented on app/commands/scrape_command.rb:
```diff
@@ -0,0 +1,47 @@
+
+  private
+
+  def persist_scrape_result(keyword, parse_result)
```
A more idionatic Ruby method name would be:

```suggestion
  def save_scrape_result(keyword, parse_result)
```

As it is consistent with the ActiveRecord API.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/defa7f47009b8c127be165fef468c28c5b83d6d9

olivierobert commented on app/commands/scrape_command.rb:
```diff
@@ -0,0 +1,47 @@
+    keyword = Keyword.find(@keyword_id)
+    html = Google::SearchService.search!(keyword.content)
+    result = Google::ParseService.new(html).call
+    persist_scrape_result keyword, result
```
Prefer to add a lin break between variable statements and method calls for readability:

```suggestion

    persist_scrape_result keyword, result
```

Make the code breathe 🌬️

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/defa7f47009b8c127be165fef468c28c5b83d6d9

olivierobert commented on app/models/keyword.rb:
```diff
@@ -0,0 +1,7 @@
+class Keyword < ApplicationRecord
+  enum status: { processing: 'processing', succeeded: 'succeeded', failed: 'failed' }
+  has_many :keyword_search_entries, dependent: :destroy
+  validates :content, :status, presence: true
```
Again here, line break issues:

```suggestion
	enum status: { processing: 'processing', succeeded: 'succeeded', failed: 'failed' }
	
  has_many :keyword_search_entries, dependent: :destroy
  
  validates :content, :status, presence: true
```

Make sure to follow this structure for all models: https://nimblehq.co/compass/development/code-conventions/ruby/ruby-on-rails/#models

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/b2aef75f8792a54884596c0bd2e63c805121f4ff

olivierobert commented on app/services/google/parse_service.rb:
```diff
@@ -0,0 +1,65 @@
+    private
+
+    def top_ads
+      @doc.css('#taw div[data-text-ad]').map do |ad_div|
```
Avoid hard-coding HTML selectors in method calls. For code sicoverabaility, it would be best to define these as constants.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits

olivierobert commented on app/services/google/parse_service.rb:
```diff
@@ -0,0 +1,65 @@
+    private
+
+    def top_ads
+      @doc.css('#taw div[data-text-ad]').map do |ad_div|
```
We can avoid using `@` to access instance variables by defining these variables as private `attr_accessor`:

```
private

attr_accessor :html, :doc
```

Then it becomes:

```suggestion
      doc.css('#taw div[data-text-ad]').map do |ad_div|
```

It is an idiomatic Ruby way 🧑‍🏫

olivierobert commented on app/services/google/parse_service.rb:
```diff
@@ -0,0 +1,65 @@
+      end
+    end
+
+    def extract_links(doc) = doc.css('a[href][data-ved]').pluck(:href)
```
Single-line methods are rarely used in Ruby (again, it is different from Elixir 😉 )

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,32 @@
+    SearchServiceError = GoogleScraperRuby::Errors::SearchServiceError
+    HOST = 'www.google.com'
+    # rubocop:disable Layout/LineLength
+    USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36'
```
To prevent being blocked when doing multiple searches, you will need to implement user agent rotation. So it is best to start planning adding this in this PR.

olivierobert replied:
Will this be addressed in this PR or another one?

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/3d1e46eaf9eb5a2136fcad7ebdf5189fb0e49c28

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,32 @@
+
+module Google
+  module SearchService
+    SearchServiceError = GoogleScraperRuby::Errors::SearchServiceError
```
This error muest be defined in: https://github.com/Goose97/google-scraper-ruby/pull/20/files#diff-abc06f343e143ef5f7301a4a974abcace033c43ef32b083a833a91e87a11def9

Goose97 replied:
well, this is just an alias, for better readability (again, inspired by Elixir)

olivierobert replied:
Ah I see, but again this is Ruby 😅 Aliases are not a thing in Ruby. So you need to follow Ruby idiomatic ways. You cannot/shouldn't code Ruby following another language's code style.

Goose97 replied:
Fixed in 86cae6015bd328076e65b0f3887b828090f28c13

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,32 @@
+      raise SearchServiceError.new uri, e
+    end
+
+    def self.build_conn(uri)
```
`conn` is something taken from Phoenix which has no place in Rails 😅 Here the method name should be:

```suggestion
    def http_client
```

Since it is not used elsewhere, it could be a private instance method in a class.

Goose97 replied:
actually, it's Faraday's terminology (https://www.rubydoc.info/github/lostisland/faraday/Faraday.new)

However, I admit that a framework-agnostic name would be better. Good point.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/2f3c3930c6f5b0f8a32041d9e8e88ae9b7e91828

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Google
+  module SearchService
```
Modules with method are rarely used. It is not idiomatic in Ruby unless for mixins. So this rather be a standard class. In Rails, it is best to follow OOP patterns, are the framework itself follows them. Other Ruby frameworks follow functional programming oriented paradigms but Rails is not one of them,

Goose97 replied:
got it. While we're at it, which one is more idiomatic:

Using class methods
```ruby
module Google
  class SearchService
    def search!(keyword)
    end
  end
end
```

Using instance methods
```ruby
module Google
  class SearchService
    def initialize(keyword)
      @keyword = keyword
    end

    def search!
    end
  end
end
```

olivierobert commented on spec/commands/scrape_command_spec.rb:
```diff
@@ -0,0 +1,45 @@
+RSpec.describe ScrapeCommand, type: :service do
+  describe '#run' do
+    context 'given an existing keyword_id' do
+      it 'updates the scrape result to the database if the command succeeds', vcr: 'google/google_no_ads' do
```
Prefern waraping conditional into a context to adoid lengthy example descriptions:

```ruby
context 'given an existing keyword' do
  context 'given NO errors' do
    it 'save the scraped result', vcr: 'google/google_no_ads' do

    end
  end

  context 'given errors' do
    it 'raises an error' do

    end

    it 'logs the error' do

    end
  end
end
```

> **Note**
> Notice the capitalization of `NO` to make it easy at first glance to know what the test is about.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/642b749eead67d6bae83f506917819c2f03c55ca

olivierobert commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,80 @@
+
+RSpec.describe Google::ParseService, type: :service do
+  describe '#call' do
+    context 'when parsing a page having 3 top ads and 1 bottom ad' do
```
The test descriptions are too coupled with the fixtures making it hard to maintain. Prefer the following structure:

```
context 'given a page with top and bottom ads' do
  it 'retrieves the top ads', vcr: 'google/programming_courses_3_top_1_bottom' do

  end

  it 'retrieves the bottom ads', vcr: 'google/programming_courses_3_top_1_bottom' do

  end
end
```

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/72b6f0e445d318a8a42cf6cddd88123223a0ef1b

olivierobert commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,80 @@
+      end
+    end
+
+    context 'when parsing a page having no ads' do
```
An improved description would be:

```suggestion
    context 'a page WITHOUT any ads' do
```

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/9988cc8d24c0bff170a67f8cc38e7b31a6adf90d

olivierobert commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,80 @@
+      end
+    end
+
+    context 'when parsing a page contains an ad for apple.com/store' do
```
These tests are duplicated with the ones for top and bottom contexts. See if these can be merged.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/72b6f0e445d318a8a42cf6cddd88123223a0ef1b#diff-a64e6d1a295f3b99057c1fa3f636d0b6ce962ebb7c5957bbda89e526ab28795bL35-L36

olivierobert commented on spec/services/google/search_service_spec.rb:
```diff
@@ -0,0 +1,59 @@
+  let(:stubbed_conn) { Faraday.new { |b| b.adapter(:test, stubs) } }
+
+  describe '#search!' do
+    before(:each) do
```
Avoid using let and before blocks as much as possible: https://nimblehq.co/compass/development/code-conventions/ruby/rspec/#convention-15. There is added complexity when nesting context with the order of execution which can cause unexpected side effects. Prefer to inline the mock in each test even if it comes at the cost of repetitions, and define helper methods instead.

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/a65b875f9864c4458482e66e048d38338ae16bad

However, this also makes some test cases exceed the spec example length limit. IMO, we can disable the lint rule in these cases

Goose97 committed:
Update gem description

Co-authored-by: Olivier Robert <olivierobert@gmail.com>

Goose97 committed:
Change method name to a Ruby idiomatic one

Goose97 committed:
Follow company style conventions regarding models

Goose97 committed:
Make attr_accessor private and avoid hard-coding constants

Goose97 committed:
Improve code style

- Use a framework-agnostic method name: build_conn -> http_client
- Use class instead of module

Goose97 committed:
Revert rubocop.yml change

Goose97 committed:
Rename from command to service

Goose97 committed:
Improve parse_service test cases

Goose97 committed:
Improve scrape_service test cases

Goose97 committed:
Improve search_service test cases

olivierobert commented on app/services/google/parse_service.rb:
```diff
@@ -0,0 +1,79 @@
+    INSIDE_SEARCH_ENTRY_LINKS_SELECTOR = 'a[href][data-ved]'
+
+    private_constant :TOP_ADS_SELECTOR, :BOTTOM_ADS_SELECTOR, :NON_ADS_NORMAL_SELECTOR, :NON_ADS_VIDEO_SELECTOR,
+                     :GENERAL_LINKS_SELECTOR, :INSIDE_SEARCH_ENTRY_LINKS_SELECTOR
```
Using 	`private_contant` is quite rare (i.e., rarely seen in Ruby codebases) 🤔  These could be moved in the public interface without any issues, but it is fine to keep as-is.

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,32 @@
+      raise SearchServiceError.new uri, e
+    end
+
+    def self.http_client(uri)
```
This does not need to be a class method:

```suggestion
    private 
    
    def http_client(uri)
```

Goose97 replied:
Fixed in 9a60d37461cdc3d701414195e55e9592dd5e84b3

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,32 @@
+    def self.search!(keyword)
+      uri = URI::HTTPS.build(host: HOST, path: '/search', query: { q: keyword }.to_query)
+      conn = http_client uri
+      response = conn.get(uri, {}, { 'User-Agent' => USER_AGENT })
```
Code style improvemens: add blank lines, rename variables to complete names and more consistency in using parentheses (you used them for `URI::HTTPS.build` but not for the line below)

```suggestion
      uri = URI::HTTPS.build(host: HOST, path: '/search', query: { q: keyword }.to_query)
      client = http_client(uri)
      response = client.get(uri, {}, { 'User-Agent' => USER_AGENT })
      
```

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,32 @@
+    USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36'
+    # rubocop:enable Layout/LineLength
+
+    def self.search!(keyword)
```
Since it is class, let's use a constructor method and an instance method to follow OOP:

```ruby
 def initialize(keyword)
   @keyword = keyword
end

def search!
   #...
end

private

attr_reader :keyword
```

Goose97 replied:
Fixed in 9a60d37461cdc3d701414195e55e9592dd5e84b3

olivierobert commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,76 @@
+    context 'given a page with top and bottom ads' do
+      it 'retrieves the top ads', vcr: 'google/programming_courses_3_top_1_bottom' do
+        html = Google::SearchService.search!('programming courses')
+        top_ads = described_class.new(html).call.search_entries.count { |e| e.kind == :ads && e.position == :top }
```
You seem to like compacting code 😅 It hurts readability for humans. Since all tests usually follow the [AAA pattern](https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80), it helps to add a line break around each block.

```ruby
html = Google::SearchService.search!('programming courses')

top_ads = described_class.new(html).call.search_entries.count { |e| e.kind == :ads && e.position == :top }

expect(top_ads).to eq(3)
```

Goose97 replied:
good catch, I just missed this one 😰

Goose97 replied:
Fixed in f4450f1df3ade108e9cc56259f081318d4caa2d3

olivierobert commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,76 @@
+        expect(bottom_ads).to eq(1)
+      end
+
+      it 'retrieves all the ads', vcr: 'google/programming_courses_3_top_1_bottom' do
```
This test is kind of redundant with the two previous tests. It does not increase confidence. See [the test matrix](https://nimblehq.co/compass/development/testing/#test-matrix) to understand more the tradeoff about confidence when it comes to testing 🧑‍🏫

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/3c9ab33336160ac9cae55cb2a98d4440075c95ee

Goose97 replied:
The initial rationale is to make sure parse service returns exclusively :top and :bottom ads. Now I think about it, it's neither a feature requirement nor this test ensures that.

olivierobert commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,76 @@
+      end
+    end
+
+    context 'given a page with non-ads entries' do
```
How is this different from the previous context? You might misunderstand the purpose of `context` as the goal is not to create blocks but identify the different branches of code. In this case, there are only two branches (page with or without ads), so only 2 contexts are required.

Goose97 replied:
I'm indeed misunderstand the meaning of the context block.

So in this case, should this test

```ruby
context 'given any page' do
  it 'includes the html of such page', vcr: 'google/monitor_50_links' do
    html = Google::SearchService.search!('monitor')
    expect(described_class.new(html).call.result_page_html).to be_a String
  end
end
```

appear in both contexts?

andyduong1920 replied:
@Goose97 for the non-ads, count the total link or the HTML page, no need to include in any context as it's the general case

```
describe '#call' do
  it 'retrieves non-ads entries',

  it 'extracts the non-ads entry URLs',

  it 'counts the number of links',

  it 'includes the html result

  context 'given a page with top and bottom ads' do
    ...
end
```

Goose97 replied:
Fixed in 99d2ec2c8842dcd2391156ebc265a4c65d733c6e

olivierobert commented on spec/services/google/scrape_service_spec.rb:
```diff
@@ -0,0 +1,84 @@
+          end.to raise_error(SearchServiceError)
+        end
+
+        it 'logs an error' do
```
@Goose97 That is fine to increase the line count to `8` or `10` as you previously wanted to do it.

olivierobert commented on spec/services/google/scrape_service_spec.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+    end
+
+    context 'given an non-existing keyword_id' do
```
English syntax issue and rewording:

```suggestion
    context 'given an INVALID keyword_id' do
```

Goose97 committed:
Randomize user agent on every search requests

olivierobert commented on spec/services/google/scrape_service_spec.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+
+      it 'logs an error' do
+        command = described_class.new(keyword_id: 42)
```
There is no more command object anymore so this needs to be renamed to match the current implementatio

Goose97 replied:
Fixed in c8f0111f6aedc64645a3299d7e6d9034e5afa174

olivierobert commented on spec/services/google/search_service_spec.rb:
```diff
@@ -0,0 +1,67 @@
+RSpec.describe Google::SearchService, type: :service do
+  describe '#search!' do
+    context 'when search a keyword' do
+      it 'perform a request against www.google.com' do
```
Again, some issues with the usage of context here. For consistency, prefer using `given` (instead of mixing `given` and `when` in the same codebase). Here there should be two contexts: `given a successful request` and `given a failed request`

Goose97 replied:
Fixed in df8db9fa717be526c4f37e066016095bc23c30e6

olivierobert commented on spec/services/google/search_service_spec.rb:
```diff
@@ -0,0 +1,67 @@
+        end
+
+        described_class.search!('hello world')
+        expect(outer_env.url.host).to include('www.google.com')
```
Many of the tests in this file test implementation details, which is a code smell. Instead, only test the behavior (the B in BDD testing). The tests should not know too much about the implementation but only care about the output of the code. It makes tests easier to maintain as we can swap the implementation while keeping the same behavior without breaking the tests. 

So to illustrate this point, there is no value in checking which host the code requests to. What matters is what the request returns. So here for simplicity we can also use `vcr` wand not use any stubs.

Goose97 replied:
I agree with your opinion regarding the target host test case

To make sure we're on the same page. This test case
```ruby
it 'raise GoogleScraperRuby::Errors::SearchServiceError if the request returns a 4xx status code' do
  HttpClientHelpers.stub_http_adapter do |adapter, connection|
    allow(described_class).to receive(:http_client).and_return(connection)
    adapter.get('/search') { [429, {}, 'Too Many Requests'] }
  end

  expect do
    described_class.search!('hello world')
  end.to raise_error(GoogleScraperRuby::Errors::SearchServiceError)
end
```

also knows a lot about SearchService implementation (the face that it calls `http_client` to build the http client). So a better solution is to use a vcr which returns a 429 response. This will make sure the test is still correct even if we change the HTTP client library. Is it correct?

Goose97 replied:
also, I'm struggling to decide if this is a behavior we want to test:

```ruby
it "raise GoogleScraperRuby::Errors::SearchServiceError if it can't connect to the host" do
  HttpClientHelpers.stub_http_adapter do |adapter, connection|
    allow(described_class).to receive(:http_client).and_return(connection)
    adapter.get('/search') { raise Faraday::ConnectionFailed }
  end

  expect do
    described_class.search!('hello world')
  end.to raise_error(GoogleScraperRuby::Errors::SearchServiceError)
end
```

If it's worth testing then how should we do it. The current implementation is leaking implementation details i.e.  the `Faraday::ConnectionFailed` error

Goose97 committed:
Improve search service error ergonomic

Goose97 committed:
Fix test description wording

Co-authored-by: Olivier Robert <olivierobert@gmail.com>

Goose97 committed:
Remove redundant tests

Goose97 committed:
Increase rspec example length limit

Goose97 committed:
Improve tests readability

Goose97 committed:
Remove command terminology for consistency

Goose97 committed:
Improve test description consistency

Goose97 committed:
Remove module name alias

Goose97 committed:
Reorganize test cases context

olivierobert commented on app/lib/google_scraper_ruby/errors.rb:
```diff
@@ -0,0 +1,33 @@
+      attr_reader :original_error, :url
+    end
+
+      def initialize(url, error = nil)
```
Seems like a rebase or copy/paste gone wrong as this code block seems to come from somewhere unexpected

Goose97 replied:
my bad. It was a rebase error, already fixed.

andyduong1920 commented on app/services/google/parse_service.rb:
```diff
@@ -0,0 +1,79 @@
+    private_constant :TOP_ADS_SELECTOR, :BOTTOM_ADS_SELECTOR, :NON_ADS_NORMAL_SELECTOR, :NON_ADS_VIDEO_SELECTOR,
+                     :GENERAL_LINKS_SELECTOR, :INSIDE_SEARCH_ENTRY_LINKS_SELECTOR
+
+    attr_accessor :doc, :html
```
How about we just allow read only?

```suggestion
    attr_reader :doc, :html
```

andyduong1920 commented on app/services/google/parse_service.rb:
```diff
@@ -0,0 +1,79 @@
+
+    def top_ads
+      doc.css(TOP_ADS_SELECTOR).map do |ad_div|
+        SearchEntry.new(
```
instead of using the SearchEntryStruct here, how about we use the ActiveModel record directly @Goose97 as we are rarely use/define raw Struct in a Rails project, mostly we will use ActiveRecord or ActiveModel for that.

So here how about we use `KeyWordSearchEntry.new(xxx)`

andyduong1920 replied:
If you plan to use the Struct here btw, I would suggest to add the Struct Suffix to make it clear

Goose97 replied:
That's a good point. I admit that there's a lot of overlapping between `KeywordSearchEntry` model and `SearchEntryStruct`. But the `KeywordSearchEntry` comes with all the baggage of `created_at`, `updated_at`, association fields, etc. I'd rather to keep these fields out of the result.

So I guess I'd keep the structs as is and add the Struct suffix as your suggestion

Goose97 replied:
Fixed in 57a134fe4ae0c6e85d29c63a25ea85e3a6ef2514

andyduong1920 replied:
> But the KeywordSearchEntry comes with all the baggage of created_at, updated_at, association fields, etc.

Is there any issue with that @Goose97 for us, we don't see any issue, also the created_at, updated_at doesn't have any value when we `new` the KeywordSearchEntry object FYI, I have tested with a model on my local

<img width="494" alt="image" src="https://github.com/Goose97/google-scraper-ruby/assets/11751745/339d6794-235a-4d92-9ee7-c23ba66b6301">

Goose97 replied:
Oh, my rationale here is: I want to keep the ParseService to detach from the database layer. Fields like `created_at`, `updated_at` or associations are database concepts, whether or not they have value. This is just my preferences, no noticeable runtime differences.

However, after weighing the pros and cons, I think it better to stick with the team conventions here.

Goose97 replied:
Fixed in bb8653ab5d4283653e4512deee39fc3758aecb96

andyduong1920 commented on app/models/keyword_search_entry.rb:
```diff
@@ -0,0 +1,10 @@
+
+class KeywordSearchEntry < ApplicationRecord
+  enum kind: { ads: 'ads', non_ads: 'non_ads' }
+  enum position: { top: 'top', bottom: 'bottom' }
```
btw we have the rightads position as well @Goose97 

![image](https://github.com/Goose97/google-scraper-ruby/assets/11751745/519fd89b-1248-41f8-8a49-1f1574adaca0)

Goose97 replied:
Somehow, I can not find a search keyword that will show these kind of ads so I can create a vcr fixture for it.

What is your search phrase?

andyduong1920 replied:
my keyword there is just `Steam deck` @Goose97 :D

Goose97 replied:
Still no ads :(. Can we leave this one for future releases?

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,50 @@
+
+module Google
+  class ScrapeService
+    attr_reader :keyword
```
Is it keyword or keyword_id 🤔 

```suggestion
    attr_reader :keyword_id
```

Goose97 replied:
Fixed in latest version

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,50 @@
+    # rubocop:disable Metrics/MethodLength
+    # rubocop:disable Metrics/AbcSize
+    def call
+      keyword = Keyword.find(@keyword_id)
```
if we defined `attr_reader :keyword_id` above, here we just use `keyword_id` without the `@`

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/b0713248018bbc87ba507fb4a5ffa388a807a3df#diff-c3acbe2de6be98fe66d47a5ec8d1965cf566a1e37807230b81b87e88c875cb27L14-L15

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,50 @@
+
+    # rubocop:disable Metrics/MethodLength
+    # rubocop:disable Metrics/AbcSize
+    def call
```
as this method could raise the exception, let follow the Ruby convention 

- use `!` suffix if the method could raise exception

https://albertoalmagro.com/three-things-you-must-know-about-ruby-bang-methods/


```suggestion
    def call!
```

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/1ef3fbf1560eb98c39c5780571d2df1442e949a2

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,50 @@
+    - keyword: #{keyword.content}
+    - error: #{e}"
+
+      raise e
```
Shall we update the `failed` status for the Keyword here?

Goose97 replied:
I intended to handle to status update part in the [background job integration ticket](https://github.com/Goose97/google-scraper-ruby/issues/15)

andyduong1920 replied:
Shall we add the TODO so the reviewer will aware about your plan @Goose97

Goose97 replied:
yeah, sure. Fixed in e64878013a57db1b0ec6884073940edd47c5a74e

andyduong1920 commented on app/services/google/parse_service.rb:
```diff
@@ -0,0 +1,79 @@
+        SearchEntry.new(
+          kind: :non_ads,
+          urls: extract_links(div),
+          position: nil
```
instead of `nil` can we use `page` or `main_search` to make it clear @Goose97 🤔

Goose97 replied:
Good call. Fixed in fe37c32bc8aa7ddc94fafbd011e07db7633f3d07

andyduong1920 commented on db/migrate/20230727072004_create_keywords.rb:
```diff
@@ -0,0 +1,26 @@
+
+    create_table :keywords, id: :serial do |t|
+      t.string :content, limit: 255, null: false
+      t.enum :status, enum_type: 'keyword_status', default: 'processing', null: false
```
should default as `pending` instead of `processing` 🤔 it will change to processing when we actual start the process 🤔

Goose97 replied:
Good point. Fixed in 5e498403fb9cb5f3583c9e5f701826ce34f60c42

Goose97 committed:
Improve scrape service error handling

- Only raise one class of error
- Better error information encapsulation

andyduong1920 commented on spec/services/google/scrape_service_spec.rb:
```diff
@@ -0,0 +1,80 @@
+        begin
+          described_class.new(keyword_id: 42).call
+        rescue ActiveRecord::RecordNotFound
+          expect(Rails.logger).to have_received(:error).with(/keyword doesn't exist/)
```
In general it's ok to write more log information but for some simple case like RecordNotFound here, I think we do not need add more content for that because the default Rails framework message pretty clear already

Goose97 replied:
I merged two errors into one class `GoogleScraperRuby::Errors::ScrapeServiceError`. Even though it has a `kind` field to distinguish between two variants, it's not as clear as with `RecordNotFound`. So, should we keep the log?

andyduong1920 replied:
In most case with the `RecordNotFound` we know that the Keyword doesn't exist :D but you can keep the log there 🙆‍♂️

andyduong1920 commented on spec/services/google/scrape_service_spec.rb:
```diff
@@ -0,0 +1,80 @@
+          keyword = Fabricate :keyword
+
+          described_class.new(keyword_id: keyword.id).call
+          updated = Keyword.includes(:keyword_search_entries).find(keyword.id)
```
```suggestion
          updated_keyword = keyword.reload
```

Goose97 replied:
Fixed in 83904a70dcb01dce9120c1bd4bceb0109c3bb2c7

andyduong1920 commented on spec/services/google/search_service_spec.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    context 'given a failed request' do
+      it 'raise GoogleScraperRuby::Errors::SearchServiceError if the request returns a 4xx status code' do
```
`it raises`

Goose97 replied:
Fixed in 6abe0de029aea0819556cb8a42fc90354de264fa

andyduong1920 commented on spec/services/google/search_service_spec.rb:
```diff
@@ -0,0 +1,55 @@
+      end
+    end
+
+    context 'given a failed request' do
```
```
context 'given a 2xx response`
  it 'xxx

context 'given a 4xx response`
  it 'raises xxx

context 'given a 5xx response`
  it 'raises xxx

context 'given a connection fail issue`
  it 'raises xxx
```

Goose97 replied:
Fixed in 6abe0de029aea0819556cb8a42fc90354de264fa

andyduong1920 commented on app/models/keyword.rb:
```diff
@@ -0,0 +1,9 @@
+
+  has_many :keyword_search_entries, dependent: :destroy
+
+  validates :content, :status, presence: true
```
should we add the validation for the `result_page_html` if the status is `succeeded` @Goose97 

```
validates :result_page_html, presence: true, if: succeeded?
```

Goose97 replied:
Nice catch. Fixed in 77490778458b6389e4af1cc565d529aa829ee7cc

andyduong1920 commented on app/lib/google_scraper_ruby/errors.rb:
```diff
@@ -0,0 +1,33 @@
+      def initialize(url:, error: nil)
+        @original_error = error
+        @url = url
+        super('')
```
Can we just send the original error to the parent 
```suggestion
        super(error)
```

Goose97 committed:
Convert SearchService class methods to instance methods

Goose97 committed:
Change attributes to read only

Co-authored-by: Andy Duong <andyduong1920@gmail.com>

Goose97 committed:
Propagate original error to parent class

Co-authored-by: Andy Duong <andyduong1920@gmail.com>

Goose97 committed:
Reorganize test cases

Goose97 committed:
Use reload to refetch model

Goose97 committed:
Reorganize test cases

Goose97 committed:
Use dependency injection for ScrapeService

Goose97 committed:
Add Struct suffix to struct names

Goose97 committed:
Add position to non_ads entry

olivierobert commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+
+    # rubocop:disable Metrics/MethodLength
+    def call!(search_service: nil)
```
[As commented on Slack](https://nimble-co.slack.com/archives/C05HAMT941Z/p1690897046668539), `search_service` should be added as a param in the constructor. Also, we can likely default to `Google::SearchService` if no param is provided:

```ruby
def initialize(keyword_id:, search_service: Google::SearchService)
```

Goose97 replied:
Fixed in 6fbb5e0caa34d9cb91ab35a3f314d40cac37eb95

olivierobert commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,72 @@
+      begin
+        search_service ||= Google::SearchService.new(keyword.content)
+        html = search_service.search!
+      rescue GoogleScraperRuby::Errors::SearchServiceError => e
```
Avoid [abbreviated variable names](https://nimblehq.co/compass/development/code-conventions/#convention-12):

```suggestion
      rescue GoogleScraperRuby::Errors::SearchServiceError => error
```

Prefer spelling out every variable name. You are not saving many keystrokes 😉  and it helps readability in the long run.

Goose97 replied:
Fixed in 19712c4f4945fdd0cb24eb04d37db0d532b43fa5

olivierobert commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,72 @@
+        )
+
+        # rubocop:disable Rails/SkipsModelValidations
+        keyword.keyword_search_entries.insert_all parse_result.search_entries.map(&:to_h)
```
As raised before, it would be better to be consistent in the usage of parenthesis. Either use them all them time or never, but not only sometimes. In this case, parenthesis is used for `map` but not `insert_all` 🙃 

```suggestion
        keyword.keyword_search_entries.insert_all(parse_result.search_entries.map(&:to_h))
```

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,59 @@
+    def search!
+      uri = URI::HTTPS.build host: HOST, path: '/search', query: { q: keyword }.to_query
+      conn = http_client uri
+      response = conn.get uri, {}
```
Let's add a line break here too:

```suggestion
      response = conn.get uri, {}
      
```

olivierobert commented on app/services/google/search_service.rb:
```diff
@@ -0,0 +1,59 @@
+    # rubocop:disable Metrics/MethodLength
+    def search!
+      uri = URI::HTTPS.build host: HOST, path: '/search', query: { q: keyword }.to_query
+      conn = http_client uri
```
It seems like you missed my previous comment about not using the abbreviated name coming from the Phoenix world 😅 

```suggestion
      client = http_client(uri)
```

Goose97 replied:
Fixed in 9b3698824c9292c74247204b22f7df5c0fbed19c

olivierobert reviewed (commented):
The tests part is much better now, but make sure not to miss any previously raised comments, to avoid lengthy back and forth.

Goose97 committed:
Consistent use of parentheses in function calls

Co-authored-by: Olivier Robert <olivierobert@gmail.com>

Goose97 committed:
Remove conn abbreviation

Goose97 committed:
More idiomatic depedency injection for ScrapeService

- Inject depedencies during instance initialization, not during call!

Goose97 committed:
Add keyword model validation rules

Goose97 committed:
Change e to error

Goose97 committed:
Minor style changes

andyduong1920 commented on app/lib/google_scraper_ruby/errors.rb:
```diff
@@ -0,0 +1,46 @@
+      attr_reader :original_error, :url
+    end
+
+    class ScrapeServiceError < StandardError
```
in Ruby we would prefer to have separate file for each class instead of having a single file for all classes @Goose97 so let's move this to 2 files 

- app/lib/google_scraper_ruby/errors/search_service_error.rb
- app/lib/google_scraper_ruby/errors/scrape_service_error.rb

I also recommend to remove the `_service_` in the file name, so it's just about the `search_error` or `scrape_error` what do you think?

Goose97 replied:
Fixed in https://github.com/Goose97/google-scraper-ruby/pull/20/commits/54cff3f50e72c4bd4d34390f93d1c25caf25b5bc

andyduong1920 replied:
@Goose97 the fix is just about rename the class, both class still in the same file, please double check my above comment 🙏

Goose97 replied:
my bad. Fixed in 8d8f1ce6d19a1481c0ac1e8b21c7c935ce4f4bd0

andyduong1920 commented on app/models/keyword.rb:
```diff
@@ -0,0 +1,20 @@
+
+  private
+
+  def succeeded?
```
with [Enum](https://api.rubyonrails.org/v7.0.6/classes/ActiveRecord/Enum.html), it will define the `succeeded?` method for us  FYI, so we can remove this one ;)

Goose97 replied:
Fixed in 3f79f1904b1d6154608ebca568f0920f040814ec

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,70 @@
+
+    def call!
+      keyword = Keyword.find_by id: keyword_id
+      raise_keyword_not_found unless keyword
```
why not just use the `Keyword.find(keyword_id)` like previous version and we catch on the RecordNotFound exception like before 🤔

Goose97 replied:
oh I think this makes it clearer where the error came from. Try to avoid potential confusion

andyduong1920 replied:
no need to do that @Goose97 as in Rails, we know that the `find(xx)` will raise the NotFound exception, so we just catch it.

Goose97 replied:
Thanks for the note. Fixed in 736b1d120535fc0713ebd5fa47eb230a634031b8

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,70 @@
+      @keyword = keyword
+
+      begin
+        html = search_service.search! keyword.content
```
how about move this into a separate method

```
html_response = search

private

def search
  search_service.search! keyword.content
rescue GoogleScraperRuby::Errors::SearchServiceError => error
        raise_unexpected_error error
end
```

Goose97 replied:
Sure 22323d5506c9b7c5b6d0f0ff979c9b97fac3df4b

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,70 @@
+        raise_unexpected_error error
+      end
+
+      result = Google::ParseService.new(html).call
```
Above we pass the `search_service` and here we use the `Google::ParseService` directly, so we keep the pattern consistency, should we pass the `parse_service` to this as well 🤔

Goose97 replied:
ah yeah, I was thinking about it as well but decide to leave it like this. No particular reason. Your point makes sense

Goose97 replied:
Fixed in b5840010db8bda70161194af39239f14661508f9

andyduong1920 commented on db/migrate/20230727072004_create_keywords.rb:
```diff
@@ -0,0 +1,26 @@
+    create_table :keyword_search_entries, id: :serial do |t|
+      t.enum :kind, enum_type: 'search_entry_kind', null: false
+      t.string :urls, array: true, default: []
+      t.enum :position, enum_type: 'search_entry_position'
```
```suggestion
      t.enum :position, enum_type: 'search_entry_position', null: false
```

andyduong1920 commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,87 @@
+    it 'counts the number of links', vcr: 'google/monitor_50_links' do
+      html = Google::SearchService.new.search! 'monitor'
+
+      count = described_class.new(html).call.links_count
```
```suggestion
      links_count = described_class.new(html).call.links_count
```

or we can just test it directly

```
expect(described_class.new(html).call.links_count).to eq(50)
```

andyduong1920 commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,87 @@
+      expect(count).to eq(50)
+    end
+
+    it 'includes the html of such page', vcr: 'google/monitor_50_links' do
```
```suggestion
    it 'includes the html of the page', vcr: 'google/monitor_50_links' do
```

andyduong1920 commented on spec/services/google/scrape_service_spec.rb:
```diff
@@ -0,0 +1,90 @@
+          described_class.new(keyword_id: keyword.id).call!
+          updated_keyword = keyword.reload
+
+          expect(updated_keyword.links_count).to be_a Integer
```
should we test with the actual value, like 1 or 2?

Goose97 replied:
If so, should we update the fixture name to reflect that?

from
```
google/google_no_ads
```

to
```
google/google_no_ads_N_links
```

andyduong1920 replied:
That is not necessary I think @Goose97 or you can test the `expect(updated_keyword.links_count).to be > 0`

Goose97 replied:
Fixed in 511e334d416ff1d2b41e9419cbd7f479575f16f2

andyduong1920 commented on spec/services/google/scrape_service_spec.rb:
```diff
@@ -0,0 +1,90 @@
+          expect(updated_keyword.links_count).to be_a Integer
+        end
+
+        it 'saves all search entires', vcr: 'google/google_no_ads' do
```
should we test it updates the status to `succeed`?

Goose97 replied:
Fixed in b702425f6dac9764fb1c4a4c391774acddc991a1

Goose97 committed:
Update spec/services/google/parse_service_spec.rb

Co-authored-by: Andy Duong <andyduong1920@gmail.com>

Goose97 committed:
Update db/migrate/20230727072004_create_keywords.rb

Co-authored-by: Andy Duong <andyduong1920@gmail.com>

Goose97 committed:
Update spec/services/google/parse_service_spec.rb

Co-authored-by: Andy Duong <andyduong1920@gmail.com>

Goose97 committed:
Remove redundant method

Goose97 committed:
Inject ParseService instance to ScrapeService to keep it consistent

Goose97 committed:
Add TODO note

Goose97 committed:
Change keyword default status to pending

Goose97 committed:
Use KeywordSearchEntry model as ParseService result

Goose97 committed:
Add test case

Goose97 committed:
Fix typo

olivierobert reviewed (approved):
One Andy's comments are resolved, this can be merged. So giving a pre-approval for not blocking you :-)

Goose97 committed:
Better error name

andyduong1920 commented on db/migrate/20230803022448_add_keyword_status_variant.rb:
```diff
@@ -0,0 +1,9 @@
+      ALTER TYPE keyword_status ADD VALUE 'pending';
+    SQL
+
+    change_column_default :keywords, :status, 'pending'
```
FYI @Goose97 as we haven't merge this migration so with this kind of new change we can also edit it directly in the original migration :D app/services/google/search_service.rb, but a new migration like this as also fine just FYI

Goose97 replied:
Thanks, good to know

andyduong1920 commented on db/schema.rb:
```diff
@@ -0,0 +1,43 @@
+  create_table "keyword_search_entries", id: :serial, force: :cascade do |t|
+    t.enum "kind", null: false, enum_type: "search_entry_kind"
+    t.string "urls", default: [], array: true
+    t.enum "position", enum_type: "search_entry_position"
```
Small thing, should we add `null: false`?

Goose97 replied:
Fixed in 7a92f96ad741fbd954383290d8ccfde4d7326772

Also merge migration files to keep it tidy

andyduong1920 commented on spec/services/google/parse_service_spec.rb:
```diff
@@ -0,0 +1,87 @@
+
+      result_page = described_class.new.call(html).result_page_html
+
+      expect(result_page).to be_a String
```
should we test the String here is not an empty string `''` 😄

Goose97 replied:
Fixed in 9db72f6cbabd519c0a1d76cf3c1a887383cf9bf2

Goose97 committed:
Split errors into different files

Goose97 committed:
Improve test case

Goose97 committed:
Tidy migrations

Goose97 committed:
Make test case more strict

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,67 @@
+      )
+    end
+
+    def save_scrape_result(keyword, parse_result)
```
```suggestion
    def save_scrape_result!(keyword, parse_result)
```

Goose97 replied:
Fixed in da5fc27e929309dee83eb89e49dc76f983eb393b

andyduong1920 commented on db/migrate/20230727072004_create_keywords.rb:
```diff
@@ -0,0 +1,26 @@
+
+    create_table :keywords, id: :serial do |t|
+      t.string :content, limit: 255, null: false
+      t.enum :status, enum_type: 'keyword_status', default: 'processing', null: false
```
```suggestion
      t.enum :status, enum_type: 'keyword_status', default: 'pending', null: false
```

Goose97 replied:
Oops, git actions gone rogue here. Fixed in 91c28f7a80275a762abfc0bb70c22db03b15c423

Goose97 committed:
Improve code style

Goose97 committed:
Fix db migration

Goose97 committed:
Bang method name

Goose97 committed:
Refactor

Goose97 committed:
Better naming

andyduong1920 commented on app/services/google/scrape_service.rb:
```diff
@@ -0,0 +1,68 @@
+
+    def call!
+      # TODO(Goose97): update failed status once we integrate with background job
+      begin
```
just a small thing, we do not need the `begin` here @Goose97 so you can delete it

andyduong1920 reviewed (approved):
Just a small comment, once it fixed we can merge this PR @Goose97 💪
